pkg/ipam/node.go

```diff
@@ -782,3 +759,79 @@ func (n *Node) syncToAPIServer() (err error) {
 
 	return err
 }
+
+// update is a helper function for syncToAPIServer(). This function updates the
+// CiliumNode resource spec or status depending on `status`. The resource is
+// updated from `origNode` to `node`.
+func (n *Node) update(
+	origNode, node *v2.CiliumNode,
+	attempts int,
+	status bool,
+) (*v2.CiliumNode, *v2.CiliumNode, error) {
+	scopedLog := n.logger()
+
+	var (
+		updatedNode    *v2.CiliumNode
+		updateErr, err error
+	)
+
+	if status {
+		updatedNode, updateErr = n.manager.k8sAPI.UpdateStatus(origNode, node)
+	} else {
+		updatedNode, updateErr = n.manager.k8sAPI.Update(origNode, node)
+	}
+
+	if updatedNode != nil && updatedNode.Name != "" {
+		node = updatedNode.DeepCopy()
+		if updateErr == nil {
+			return origNode, node, nil
+		}
+	} else if updateErr != nil {
+		scopedLog.WithError(updateErr).WithFields(logrus.Fields{
+			logfields.Attempt: attempts,
+		}).Warning("Failed to update CiliumNode spec")
+
+		node, err = n.manager.k8sAPI.Get(node.Name)
+		if err != nil {
+			return origNode, node, err
+		}
+
+		// Propagate the error in case we exit the loop without succeeding in
+		// updating the resource.
+		err = updateErr
+		node = node.DeepCopy()
+		origNode = node.DeepCopy()
+	} else /* updateErr == nil */ {
+		err = updateErr
+	}
+
+	return origNode, node, err
+}
+
+// adjustPreAllocateIfNeeded adjusts IPAM values depending on the instance
+// type. This is needed when the instance type is on the smaller side which
+// requires us to lower the PreAllocate value and include eth0 as an ENI device
+// because the instance type does not allow for additional ENIs to be attached
+// (hence limited).
+//
+// For now, this function is only needed in ENI IPAM mode. In the future, this
+// may also be needed for Azure IPAM as well.
+func adjustPreAllocateIfNeeded(node *v2.CiliumNode) {
```

Comment:
Reviewers: This function has duplicated logic (albeit small) with function `determineFirstInterfaceIndex` (seen further below). I would be open for suggestions on how to deduplicate the logic (if it seems necessary to people). The reason I didn't do it from the first time is because I couldn't think of a good place to put it besides the `eni` package. On top of that, it doesn't seem worth putting such small logic in a large package because when it would be imported here, it would be the only function used out of a large package.